Hibernate Reactive 2.0.0.Alpha2 & re-enabling all HR modules

Sanne opened the pull request:
Putting this here as it's *almost* useable.

## Versions:

1. Hibernate Reactive 2.0 doesn't exist yet - it's not quite ready, this is using 2.0.0.Alpha1 and there's some caveats attached (to be documented, postponing that to after seeing a green CI build so I get a better picture myself).
2. It's requiring Hibernate ORM 6.2.0.CR3 - ideally we actually need more patches now included in `6.2` branch of ORM, but we don't want to wait for another ORM core tag.

## Known working:

1. Primary integration tests are working fine
2. DB2 tests work fine
3. PostgreSQL tests work fine
4. MySQL tests work fine

## Known problems:

1. Panache Reactive tests : a single failure, but an odd one.
2. `Hibernate Reactive REST data with Panache` integration tests trigger some of the not-yet-supported queries (it appears to be a lazy initialization load triggered during query result processing)

## Unknown:

There's a number of other modules depending on Hibernate Reactive which I'm unable to test because of miscellaneous build & container issues. e.g. WebAuthn, Kafka, gRPC, OpenTelemetry Reactive, Smallrye Reactive Messaging.

Also, we don't have Oracle / Hibernate Reactive integration tests apparently.

cc/ @geoand @gsmet @yrodiere @cescoffier 

Fixes #31242

Sanne committed:
Revert "Disable Hibernate Reactive modules in native-tests.json"

This reverts commit bbed9a7ba51be3355481a102b643dcbe6401bb3a.

Sanne committed:
Revert "Disable the Hibernate Reactive modules"

This reverts commit f0f2a36e70612986119530e6e9127056dcbb8205.

Sanne committed:
Adapt the Hibernate Reactive bootstrap code to Hibernate ORM core 6.2

Sanne committed:
Switch to Hibernate Reactive 2.0.0.Alpha2

Sanne committed:
Adapt the Hibernate Reactive extension to Hibernate Reactive 2.0

Sanne committed:
Switch to Jakarta configuration properties

Sanne committed:
Prevent warnings about using an unexpected PostgreSQL JDBC driver

Sanne committed:
Fix compilation errors in Panache Reactive's CommonPanacheQueryImpl

Sanne committed:
Fix compilation errors in integration tests: hibernate-reactive-panache-kotlin

Sanne committed:
Adapt Hibernate Reactive extension to new Services refactoring in ORM CR3

Sanne committed:
Update expectations of integration tests

Sanne committed:
New services introduced in Hibernate ORM 6.2.0.CR3

Sanne commented:
Rebased & upgraded to Hibernate ORM 6.2.0.CR3

Sanne commented:
Rebasing to test with the other two ORM related PRs

Sanne commented:
Status update: we seem to have a setback in the Panache integration tests; I didn't have that many failures just 3 days ago.

Unfortunately the Panache ITs are of the old style: runs a series of operations and if all of them succeeds, returns "OK" - we only test for OK to be returned. On top of this, they are chained & ordered, potentially leading to cascading failures of assumptions; not great for evaluating in short time what's going on.

I didn't have time to check that so there's now many more Panache tests disabled - this to allow getting a picture of other components by CI, while we try having a better look at Panache tests.

cc/ @FroMage perhaps you can help dissecting / sorting them? It's very likely that the problem is in HR, but I suspect there's many tests failing for the same underlying issue - would be great to figure that out.

gsmet commented on extensions/panache/hibernate-reactive-rest-data-panache/deployment/src/test/resources/import.sql:
```diff
@@ -1,7 +1,7 @@
-
+insert into item(id, name, collection_id) values (1, 'first', 'full');
+insert into item(id, name, collection_id) values (2, 'second', 'full');
+alter sequence Item_SEQ restart with 3;
```
So we don't need to do the `+ 50` here? /cc @yrodiere 

Asking because it might work with the current tests but might be a problem if we add more, and it will probably come as a surprise. Asking for all the sequence restarts in the PR.

Sanne replied:
You don't need the +50 here - it's fine.

yrodiere replied:
Meaning HR doesn't support optimizers, or we don't `persist` in these tests, or we don't care about negative/conflicting IDs here, or you looked into these tests and we don't get negative IDs for some reason?

Sanne replied:
We discussed this extensively on zulip the other day?

The problem of the "+50" only manifests if your sequence increment isn't matching what Hibernate ORM expects - essentially if you're running an ORM6 application with default sequence generators, on a schema which was created for ORM 5 (and not updated to match).

But since ORM generated the schema, it's going to increment correctly:
```
Hibernate: 
    create sequence EmptyListItem_SEQ start with 1 increment by 50
Hibernate: 
    create sequence Item_SEQ start with 1 increment by 50
```

If you test it, you'll see the next insert will have id=4.

Sanne replied:
> If you test it, you'll see the next insert will have id=4.

My bad, should be id=3 I think

yrodiere replied:
> The problem of the "+50" only manifests if your sequence increment isn't matching what Hibernate ORM expects

That's not what I saw in my tests.

Quoting myself from [our discussion](https://hibernate.zulipchat.com/#narrow/stream/132094-hibernate-orm-dev/topic/Default.20increment.20size.20for.20sequences/near/339301201):

```quote
 Even if you update the increment size, if you don't initialize the sequence to the correct value, ORM 6 will start reusing old IDs.
```

> If you test it, you'll see the next insert will have id=4.

That's not consistent with the behavior of the `pooled` optimizer, which upon retrieving a sequence value of `3`, is supposed to use a pool of values going from `4-50+1=-45 to `4`. That's just how it's supposed to work.

The only ways I could see this working is if for some reason:

* Hibernate ORM deals with this problem in creative ways.
* Hibernate Reactive does not use the `pooled` optimizer.
* `alter sequence Item_SEQ restart with 3;` was actually setting the "previous" value of a sequence, so that the next increment actually returns `53`. That's not the behavior of PostgreSQL nor MariaDB as far as I can see.
* OR the sequence is incremented once before the first call to `persist` but the resulting value is never used.

Anyway, I guess neither of us has time to investigate more.

> But since ORM generated the schema, it's going to increment correctly:

The core of the problem here is not the schema, but the the next value of the sequence and how Hibernate ORM (Reactive) is going to interpret it.

Sanne replied:
Yes you're right but it's using a different generator here. It's all setup explicitly on the model.

Sanne replied:
so sorry, my previous comment was wrong as I actually had a different model in front of me. Ignore me :)

gsmet reviewed (dismissed):
In #31242, I listed what should be reverted and I would prefer we don't revert [fd56088](https://github.com/quarkusio/quarkus/pull/31454/commits/fd5608811f87a92714c56a0c779911c2b8ecf548) .
It wasn't in the list for a good reason as it reduces our dependency to the Hibernate Reactive modules in the codebase. It's a small step forward but still.

I also added a question regarding the sequences but I will let @yrodiere confirm if it's an issue or not.

gsmet commented:
If we have significant issues that are remaining, I would advise to release Alpha 5 tomorrow without Hibernate Reactive and release an Alpha6 next week when things are more stable.

I will let you decide what you want to do. To be in Alpha5, things have to be merged in `main` by this evening 7PM Paris time.

Sanne commented:
> In #31242, I listed what should be reverted and I would prefer we don't revert [fd56088](https://github.com/quarkusio/quarkus/pull/31454/commits/fd5608811f87a92714c56a0c779911c2b8ecf548) . It wasn't in the list for a good reason as it reduces our dependency to the Hibernate Reactive modules in the codebase. It's a small step forward but still.

Thanks, yes that's very nice - removed it (and rebased).

FroMage commented:
I'm looking at the Panache failures.

FroMage committed:
Fixes for HR2 in HR/Panache IT tests

gsmet reviewed (commented):
@yrodiere I will let you follow the progress of it and merge it when ready.

yrodiere reviewed (approved):
LGTM as long as we actually follow-up on all those disabled tests.

Test failures seem unrelated.

@Sanne is this good to merge?

yrodiere commented:
Switching to draft again as there's no rush anymore and there are a few tests that we'd rather fix before merging.
See https://hibernate.zulipchat.com/#narrow/stream/205413-hibernate-reactive-dev/topic/Hibernate.20Reactive.202.2E0.20.2F.20Quarkus.20integration/near/340376117

Sanne commented:
Let's make this a draft again - it was rushed to try making the Alpha tag of yesterday, but since we have more time we can try to re-enable some of those tests I had to disable first.

Among others, I hear @FroMage already fixed some of the Panache ITs so it would be nice to include those fixes rather than later reverting the fact we disabled some tests.

Sanne commented on extensions/panache/hibernate-reactive-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/reactive/rest/data/panache/deployment/AbstractGetMethodTest.java:
```diff
@@ -78,6 +79,7 @@ void shouldGetComplexHalObject() {
     }
 
     @Test
+    @Disabled("Requires fixes in Hibernate Reactive")
```
@yrodiere thanks for the offer but it's not worth your time - I'm reverting most of these as we're making progress with HR snapshots.

If there's some left when we're done, then yes we'll treat those as follow-up issues that need to be tracked.

Sanne commented:
Just a notice to everyone helping about how I'm proceeding:

in the current PR we had several tests disabled, or temporary workarounds applied. I've now switched to using HR (and ORM) snapshots, and as we make progress on HR I'm removing workarounds; and force-pushing updates to the draft.

This implies the final PR will be simpler than the original one; several workarounds that Stef had to suggest for Panache ITs have already been removed.

(This is the same approach I've been following with the ORM6 blocking branch, initial drafts were a mess, the final one was rather straight forward)

Sanne committed:
Adapt the HR extension after recent changes

Sanne committed:
Requires investigation: disable failing tests in hibernate-reactive-rest-data-panache : OpenApiIntegrationTest

Sanne committed:
Reconfigure kubernetes-service-binding-reactive integration tests for consistent configuration with other modules

Sanne committed:
Service renamed for Hibernate Reactive: ParameterMarkerStrategy

Sanne commented:
@FroMage  I'm going to need your help with the failing test `extensions/panache/hibernate-reactive-rest-data-panache/deployment/src/test/java/io/quarkus/hibernate/reactive/rest/data/panache/deployment/entity/PanacheEntityResourceMethodListenerTest.java`

After debugging the whole day and failing to being able to reproduce it in a plain Hibernate Reactive unit test, I seem to have figured that the single request the test is sending to Quarkus is being handled by [? black box of generated code] to then issue two "parallel" `persist` operations, which are triggering interleaving events on the same stream.

At least, that's my conclusion as I see method `Uni<Void> persist(Object entity)` on `MutinySessionImpl` being hit *twice*, in quick succession, and with the same instance of *entity* being passed.

I have no idea what components are in between there, and since it's reactive code stacktraces won't help - could you have a look? I suppose there's some misuse of chaining upstream.

N.B. this could have worked by chance in the older Hibernate Reactive, it's less so forgiving now.

FroMage committed:
Reworked how rest-data-panache listeners work on after listeners

Use uni.invoke() instead of subscribe because that would cause a double
subscription, so running the actions twice, in HR

Sanne committed:
Fix integration tests: hibernate-reactive-panache-kotlin

Sanne committed:
Upgrade to Hibernate ORM 6.2.0.CR4

Sanne commented:
Rebased and cleaned up history - now not using snapshots anymore, so I'll undraft.

geoand commented:
Something is wrong with those LRA tests I've seen them fail in multiple PRs

gsmet commented:
From what I could see, the Docker images are not available anymore.

gsmet commented:
I'm on it.

geoand commented:
👌

gsmet commented:
Hopefully https://github.com/quarkusio/quarkus/pull/31959 will fix it.

Sanne commented:
Rebased on #31953 , which is in turn rebased on #31959
